### PR TITLE
Emit [module: RefSafetyRules] attribute for all modules compiled with updated escape rules

### LINKED ISSUE #63692: Emit [module: RefSafetyRules] attribute for all modules compiled with updated escape rules
A `[module: RefSafetyRules(11)]` attribute is emitted to modules compiled with `-langversion:11` or higher, or compiled with .NET 7 or higher _if the module contains a `ref struct` type declaration or a method that includes a `ref` or `ref struct` parameter_.

We should remove the requirement that the module contains certain members and emit the attribute based on language version and framework only.

### PR DESCRIPTION
**Pull request by cston**
Change `SourceModuleSymbol.RequiresRefSafetyRulesAttribute()` to return `true` if the module is compiled with `-langversion:11+`, or compiled with .NET 7+, and if `-features:noRefSafetyRulesAttribute` is unset.

The rest of the change is to update tests, in most cases to use `-features:noRefSafetyRulesAttribute`.

Fixes #63692

### CODE REVIEW

**Comment by cston** [comment]
@dotnet/roslyn-compiler, please review, thanks.

The code change is small change to the compiler plus simple updates to many tests.

**Comment by cston** [comment]
@dotnet/roslyn-compiler, please review, thanks.

File: src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/LocalSlotMappingTests.cs [hunk 1]
```diff
@@ -841,7 +841,7 @@ static async Task M()
         await using IAsyncDisposable <N:1>v = F()</N:1>, <N:0>u = F()</N:0>;
     }
 }");
-            var asyncStreamsTree = Parse(AsyncStreamsTypes);
+            var asyncStreamsTree = Parse(AsyncStreamsTypes, options: TestOptions.Regular.WithNoRefSafetyRulesAttribute());
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Emit/Emit/EditAndContinue/LocalSlotMappingTests.cs:844]: These two tests look very similar but in one we directly specify the options and in the other we copy the options from `source0`. Is there a reason for that?

File: src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs [hunk 2]
```diff
@@ -2122,15 +2122,16 @@ public class PublicClass
     internal event System.Action InternalEvent;
 }
 ";
+            var parseOptions = TestOptions.Regular.WithNoRefSafetyRulesAttribute();
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Emit/Emit/CompilationEmitTests.cs:2125]: nit: consider moving this down closer to where it is used (e.g. right before `var realImage` is declared)

**Review by RikkiGibson — APPROVED** [review]

**Comment by cston** [comment]
@dotnet/roslyn-compiler, please review, thanks.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceModuleSymbol.cs [hunk 3]
```diff
@@ -539,77 +539,15 @@ internal bool RequiresRefSafetyRulesAttribute()
...
-                }
-                return method.Parameters.Any(p => parameterIncludesRef(p));
-            }
-
-            static bool parameterIncludesRef(ParameterSymbol parameter)
-            {
-                return parameter.RefKind != RefKind.None ||
-                    parameter.Type.IsRefLikeType;
+                var options = (CSharpParseOptions?)compilation.SyntaxTrees.FirstOrDefault()?.Options;
+                return options?.Features?.ContainsKey("noRefSafetyRulesAttribute") == true;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceModuleSymbol.cs:550]: nit: do we need any kind of documentation for this feature flag, or will it just be explained on a case-by-case basis?

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 4)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceModuleSymbol.cs (hunk 3 above)
Comment (cston) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceModuleSymbol.cs:550]: As we discussed offline, `-features:noRefSafetyRulesAttribute` will disable emitting `[module: RefSafetyRules]` unconditionally, and is only intended for use in tests. It is not intended for use outside of tests.

**Pull request merged**